### DISCUSSION TITLE
Where are files stored when building micropython and freezing modules?

### DISCUSSION THREAD

**Discussion by maxi07** [micropython/micropython#12094] [RP2040 / Pico] [answered]
Hello! I am trying to freeze my project into micropython. This is part of my project structure:
```
.
├── main.py
├── config.json
├── ...
├── css/
│   └── styles.css
└── modules/
    ├── module1.py
    └── module2.py
```

the modules are copied fine, but my main.py imports the config.json. This now fails with `OSError: [Errno 2] ENOENT` when freezing the files. Where would that config.json now be accessible when being frozen into the firmware?

**Comment by robert-hh** [comment]
When freezing, no file system is created. Only .py files are frozen in to the firmware image and being stored in a special data structure. The json file and other file types are ignored.

**Reply by maxi07** [reply]
this is interesting! is there any way while building the custom firmware to also include those other files? I am trying to have a complete firmware including my code to simply copy the .uf2 and everything runs as expected.

**Reply by dlech** [reply]
You will have to turn them into Python files in order to freeze them.

**Reply by maxi07** [reply]
you mean turning`config.json` to `config.json.py`? Can you please elaborate a little more? I am not sure if I understand this correctly.

**Reply by dlech** [reply]
It depends on how you are using the data, but you could put the json in a string or convert it to a Python data structure. If you still want to keep the actual .json file in the source code, you could create a Python script to automate the conversion.

I.e. like this (a string)...

```python
CONFIG_JSON = """
{
    "option1": "value1",
    "option2": "value2"
}
"""
```

or like this (a Python dict/list/etc.)...

```python
CONFIG = {
    "option1": "value1",
    "option2": "value2",
}
```

**Reply by maxi07** [reply]
While this option could work with the simple json, I have bigger HTML and CSS files which I would like to keep seperate for easier development. 
I still don't quite understand what you mean by conversion?

**Reply by dlech** [reply]
I mean you can write a Python script that runs on your computer at compile time that reads all of the .css/.html/.json files and creates a new .py file and writes all of that information to the new .py file in a format that is valid Python syntax.

**Reply by karfas** [reply]
You can embed your file content (like @dlech suggested) in .py files and create some (default) config/html/CSS files out of these frozen files, if they don't exist in the filesystem.
Just use your imagination - your program can do anything you want.

**Reply by Josverl** [reply]
Compiling to bytecode is not the only way to freeze. 
It may be an option to use `freeze_as_str` in your manifest.py

`freeze_as_str("foo.html")`

https://github.com/micropython/micropython/blob/cfcce4b5311c7b112d3628392c8012b4c29bdd63/tools/makemanifest.py#L181

Not sure how to reference this runtime though.

Looks like they can be referenced as strings, but hard to test on a phone.

https://github.com/micropython/micropython/blob/cfcce4b5311c7b112d3628392c8012b4c29bdd63/tools/makemanifest.py#L243-L252

**Comment by bixb922** [comment, marked as answer]
I wrote this utility program that I use to freeze html, css, json, image, etc files to a .py file. I add these .py files to the manifest.py to be frozen in the MicroPython image. On import, the file structure gets mounted as a read only file system on the microcontroller. For read-write files, the file structure can get deployed (copied) to the controllers flash. I have done testing and it's in active use, but probably more testing may be necessary.   If you use it, I'd be very happy to get feedback. Here is the link: https://github.com/bixb922/freezeFS/tree/main

**Reply by maxi07** [reply]
this looks promising, I will definitely check it out this weekend!

**Reply by maxi07** [reply]
this works perfectly fine for me, thank you so much!

**Comment by jimmo** [comment]
Peter wrote a script for doing this: https://github.com/peterhinch/micropython-samples/blob/master/data_to_py/data_to_py.py
